### In this notebook, we will use a very simple neural network to make predictions.

#### Ref : https://www.youtube.com/watch?v=_Z9TRANg4c0

In [1]:
# Let's start by importing some useful libraries

import tensorflow as tf
import numpy as np
from tensorflow import keras

In [2]:
# Let's now create the simplest possible neural network i.e., with 1 layer and 1 neuron

model = tf.keras.Sequential([keras.layers.Dense(units = 1, input_shape = [1])])

In [3]:
# Let's compile our model by specifying an optimizer and loss function 

model.compile(optimizer = 'sgd', loss = 'mean_squared_error')

In [4]:
# Time to provide the data to our NN

X = np.array([-1,0,1,2,3,4,5,6,7,8,9,10], dtype = float)
Y = np.array([-10,0,10,20,30,40,50,60,70,80,90,100], dtype = float)


In [5]:
# Training the NN

model.fit(X,Y,epochs = 500)

Train on 12 samples
Epoch 1/500
12/12 [==============================] - 0s 28ms/sample - loss: 3389.6296
Epoch 2/500
12/12 [==============================] - 0s 177us/sample - loss: 401.6086
Epoch 3/500
12/12 [==============================] - 0s 143us/sample - loss: 48.1483
Epoch 4/500
12/12 [==============================] - 0s 224us/sample - loss: 6.3283
Epoch 5/500
12/12 [==============================] - 0s 174us/sample - loss: 1.3723
Epoch 6/500
12/12 [==============================] - 0s 136us/sample - loss: 0.7770
Epoch 7/500
12/12 [==============================] - 0s 161us/sample - loss: 0.6977
Epoch 8/500
12/12 [==============================] - 0s 156us/sample - loss: 0.6795
Epoch 9/500
12/12 [==============================] - 0s 165us/sample - loss: 0.6688
Epoch 10/500
12/12 [==============================] - 0s 163us/sample - loss: 0.6590
Epoch 11/500
12/12 [==============================] - 0s 148us/sample - loss: 0.6494
Epoch 12/500
12/12 [=============================

In [6]:
# So now we have a model that is trained to learn the relationship between X and Y. Let's try to predict the value of 
# X which the model have not seen before

print(model.predict([25.0]))

[[249.90572]]


#### We were expecting Y to be 250 but out NN gave us the value closeby to 250. This is because NN deals with probabilities. With only 12 data points, our model calculated that there is a very high probability that the relationship between X and Y is Y = 10X. As a result, the prediction for 25 is very close t0 250 but not exactly 250.  